chemprop/uncertainty/calibrator.py

```diff
@@ -68,13 +70,59 @@ def apply_calibration(self, preds, uncs) -> Tensor:
 
 @UncertaintyCalibratorRegistry.register("platt")
 class PlattCalibrator(UncertaintyCalibrator):
-    def calibrate(self, preds, uncs, targets, mask) -> Tensor:
-        ...
-        return
-
-    def apply_calibration(self, preds, uncs) -> Tensor:
-        ...
-        return
+    """
+    A calibration method for classification datasets based on the Platt scaling algorithm.
+    As discussed in https://arxiv.org/abs/1706.04599 and Platt, J. (2000). Probabilistic outputs for
+    support vector machines and comparison to regularized likelihood methods. In A. Smola, P. 
+    Bartlett, B. Schölkopf, & D. Schuurmans (Eds.), Advances in large margin classifiers. Cambridge:
+    MIT Press.
+    In Platt's paper, he suggests using the number of positive and negative examples in the dataset
+    used to train the model to adjust the value of target probabilities used to fit the parameters. 
+    """
+
+    def calibrate(self, preds, targets, mask, training_targets: None | Tensor = None) -> Tensor:
+        if not torch.logical_or((targets[mask] == 0), (targets[mask] == 1)).all():
+            raise ValueError("Platt scaling is only implemented for binary classification tasks.")
+
+        if training_targets is not None:
+            negative_targets = (1 / ((training_targets == 0).sum(dim=0) + 2)).expand_as(targets)
+            positive_targets = (
+                ((training_targets == 1).sum(dim=0) + 1) / ((training_targets == 1).sum(dim=0) + 2)
+            ).expand_as(targets)
+
+            targets = targets.float()
+            targets[targets == 0] = negative_targets[targets == 0]
+            targets[targets == 1] = positive_targets[targets == 1]
+
+        platt_a = []
+        platt_b = []
+        for i in range(preds.shape[1]):
+            task_mask = mask[:, i]
+            t_preds = preds[:, i][task_mask]
+            t_targets = targets[:, i][task_mask]
+
+            # if is_atom_bond_targets: # Not yet implemented
+
+            def objective(parameters):
+                a = parameters[0]
+                b = parameters[1]
+                scaled_preds = torch.sigmoid(a * torch.logit(t_preds) + b)
+                nll = -1 * torch.sum(
+                    t_targets * torch.log(scaled_preds)
+                    + (1 - t_targets) * torch.log(1 - scaled_preds)
+                )
+                return nll
```

Comment:
open question (I don't know), does `scipy.optimize.fmin` play nicely with tensor inputs? If not, we might want to cast everything to numpy arrays.

Reply:
The input is a list and I changed the output to be a scaler. Tensors are still used internally for calculations but because fmin never sees the tensors, I think we should be okay.

Reply:
I believe `scipy` uses `numpy` internally, which the unpacking gets `float`s out of. I don't see why we would want to implement this going between `torch` and `numpy`. This function should just be changed to use `numpy` internally.

Reply:
I think what @davidegraff mentioned is the implementation like below. Avoiding the transformation might be more efficient, but then we need to write the function of sigmoid and logit explicitly, which hurts the readability.
```
def objective(parameters):
    a = parameters[0]
    b = parameters[1]
    
    logit_preds = np.log(t_preds / (1 - t_preds))  # Logit function
    scaled_preds = 1 / (1 + np.exp(-(a * logit_preds + b)))  # Sigmoid function
    
    nll = -np.sum(
        t_targets * np.log(scaled_preds)
        + (1 - t_targets) * np.log(1 - scaled_preds)
    )
    
    return nll
```

Reply:
use [`scipy.special.logit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.logit.html) and [`scipy.special.expit`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.expit.html), respectively.